File: EPOS/LabWindowsCVI/Example LabWindowsCVI 2010/Demo_WinDLL.c
#include <ansi_c.h>
#include <userint.h>
#include "Demo_WinDLL.h"
#include "Definitions.h"

static int EPOS_Demo;

BOOL m_oImmediately;
BOOL m_oInitialisation;
BOOL m_oMovement;
BOOL m_oUpdateActive;
BYTE m_uMode;
HANDLE m_KeyHandle;
long m_lPositionActual;
long m_lPositionStart;
long m_lPositionTarget;
unsigned int m_dErrorCode;
unsigned int m_dProfileAcceleration;
unsigned int m_dProfileDeceleration;
unsigned int m_dProfileVelocity;
WORD m_wNodeId;

/************************************************************************
Init variables with default values
************************************************************************/
void Init()
{
	m_KeyHandle = 0;

	m_lPositionStart = 0;
	m_lPositionTarget = 2000;

	m_oImmediately = 1;
	m_oMovement = 0;
	m_oUpdateActive = 0;
}

/************************************************************************
Shows a message box with error description of the input error code
************************************************************************/
int ShowErrorInformation(unsigned int dErrorCode)
{
	char* pStrErrorInfo;

	if((pStrErrorInfo = (char*)malloc(100)) == NULL)
	{
		MessagePopup("System Error","Not enough memory to allocate buffer\n");

		return 0;
	}

	if(VCS_GetErrorInfo(dErrorCode,pStrErrorInfo,100))
	{
		MessagePopup("Error Information",pStrErrorInfo);

		free(pStrErrorInfo);

		return 1;
	}
	else
	{
		free(pStrErrorInfo);
		MessagePopup("Information","Error information can't be read!");

		return 0;
	}
}

/************************************************************************
Opens communication interface via the OpenDeviceDlg
************************************************************************/
int OpenDevice()
{
	BOOL oSame = 0;
	char* pStrProtocolStackName;
	HANDLE* pKeyHandle;

	Init();

	//Allocate memory for string
	if((pStrProtocolStackName = (char*)malloc(100)) == NULL)
	{
		MessagePopup("System Error","Not enough memory to allocate buffer\n");
		exit(1);
	}

	m_KeyHandle = VCS_OpenDeviceDlg(&m_dErrorCode);
	if(m_KeyHandle)
	{
		//Clear Error History
		if(VCS_ClearFault(m_KeyHandle,m_wNodeId,&m_dErrorCode))
		{
			//Read Operation Mode
			if(VCS_GetOperationMode(m_KeyHandle,m_wNodeId,&m_uMode,&m_dErrorCode))
			{
				//Read Position Profile Objects
				if(VCS_GetPositionProfile(m_KeyHandle,m_wNodeId,&m_dProfileVelocity,&m_dProfileAcceleration,&m_dProfileDeceleration,&m_dErrorCode))
				{
					//Write Profile Position Mode
					if(VCS_SetOperationMode(m_KeyHandle,m_wNodeId,0x01/*Profile Position Mode*/,&m_dErrorCode))
					{
						//Write Profile Position Objects
						if(VCS_SetPositionProfile(m_KeyHandle,m_wNodeId,100,1000,1000,&m_dErrorCode))
						{
							//Read Actual Position
							if(VCS_GetPositionIs(m_KeyHandle,m_wNodeId,&m_lPositionStart,&m_dErrorCode))
							{
								return 1;
							}
						}
					}
				}
			}
		}

		ShowErrorInformation(m_dErrorCode);
	}
	else
	{
		MessagePopup("Open Interface","Can't open device!");
		free(pStrProtocolStackName);
	}

	return 0;
}

/************************************************************************
Main Program
************************************************************************/
int main (int argc, char *argv[])
{
	if (InitCVIRTE (0, argv, 0) == 0)
		return -1;	/* out of memory */
	if ((EPOS_Demo = LoadPanel (0, "Demo_WinDLL.uir", PANEL)) < 0)
		return -1;

	if(OpenDevice())
	{
		m_oInitialisation = 1;
		m_oUpdateActive = 1;
	}
	else
	{
		m_oInitialisation = 0;
	}

	DisplayPanel (EPOS_Demo);
	RunUserInterface ();
	DiscardPanel (EPOS_Demo);
	return 0;
}

/************************************************************************
Sets new communication settings via button 'Device Settings'
************************************************************************/
int CVICALLBACK DeviceSettings (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
			if(OpenDevice())
			{
				m_oInitialisation = 1;
				m_oUpdateActive = 1;
			}
			break;
	}
	return 0;
}

/************************************************************************
Stops the movement
************************************************************************/
int CVICALLBACK Halt (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
			if(!VCS_HaltPositionMovement(m_KeyHandle,m_wNodeId,&m_dErrorCode))
			{
				ShowErrorInformation(m_dErrorCode);
			}
			break;
	}
	return 0;
}

/************************************************************************
Starts the movement
************************************************************************/
int CVICALLBACK Move (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
 			if(!VCS_GetPositionIs(m_KeyHandle,m_wNodeId,&m_lPositionStart,&m_dErrorCode))
			{
				MessagePopup("Button Move ...", "Function GetPositionIs failed!");
			}

			if(!VCS_MoveToPosition(m_KeyHandle,m_wNodeId,m_lPositionTarget,m_oMovement,m_oImmediately,&m_dErrorCode))
			{
				MessagePopup("Button Move ...", "Function 'MoveToPosition' failed!"); 				
			}

			break;
	}
	return 0;
}

/************************************************************************
Sets device to disable state
************************************************************************/
int CVICALLBACK Disable (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	BOOL oResult;

	switch (event)
	{
		case EVENT_COMMIT:
			oResult = VCS_SetDisableState(m_KeyHandle,m_wNodeId,&m_dErrorCode);
			break;
	}
	return 0;
}

/************************************************************************
Sets device to enable state
************************************************************************/
int CVICALLBACK Enable (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	BOOL oResult;

	switch (event)
	{
		case EVENT_COMMIT:
			oResult = VCS_SetEnableState(m_KeyHandle,m_wNodeId,&m_dErrorCode);
			break;
	}
	return 0;
}

/************************************************************************
Stops timer. Status will be displayed as disconnected
************************************************************************/
void StopTimer()
{
	m_oUpdateActive = 0;
}

/************************************************************************
Restores the old settings, stops the timer and delete the key handle
************************************************************************/
int RestoreEPOS()
{
	if(m_oInitialisation)
	{
		VCS_SetDisableState(m_KeyHandle,m_wNodeId,&m_dErrorCode);
		VCS_SetOperationMode(m_KeyHandle,m_wNodeId,m_uMode/*Old Mode*/,&m_dErrorCode);
		VCS_SetPositionProfile(m_KeyHandle,m_wNodeId,m_dProfileVelocity,m_dProfileAcceleration,m_dProfileDeceleration,&m_dErrorCode);
	}

	StopTimer();

	VCS_CloseAllDevices(&m_dErrorCode);

	return 1;
}

/************************************************************************
Closes the dialog
************************************************************************/
int CVICALLBACK OkCallback (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	BOOL oResult;

	switch (event)
	{
		case EVENT_COMMIT:

			RestoreEPOS();

			QuitUserInterface (0);
			break;
	}
	return 0;
}

/************************************************************************
Sets state to relative movement
************************************************************************/
int CVICALLBACK RadioMove1 (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
			m_oMovement = 0;
			break;
	}
	return 0;
}

/************************************************************************
Sets state to absolute movement
************************************************************************/
int CVICALLBACK RadioMove2 (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
			m_oMovement = 1;
			break;
	}
	return 0;
}

/************************************************************************
Updates the display
************************************************************************/
int UpdateStatus(int panel)
{
	BOOL oEnable = 1;
	BOOL oResult = m_oUpdateActive;
	char* strActiveMode;

	if(m_oMovement != 0)
	{
		SetCtrlVal(panel,PANEL_RADIOBUTTON_1,0);
		SetCtrlVal(panel,PANEL_RADIOBUTTON_2,1);
		SetCtrlAttribute(panel,PANEL_ButtonMove,ATTR_LABEL_TEXT,"__Move Absolute");
	}
	else
	{
		SetCtrlVal(panel,PANEL_RADIOBUTTON_1,1);
		SetCtrlVal(panel,PANEL_RADIOBUTTON_2,0);
		SetCtrlAttribute(panel,PANEL_ButtonMove,ATTR_LABEL_TEXT,"__Move Relative");
	}

	if(oResult)
	{
		oResult = VCS_GetOperationMode(m_KeyHandle,m_wNodeId,&m_uMode,&m_dErrorCode);
		if(oResult)
		{
			switch(m_uMode)
			{
				case -6: strActiveMode = "Step/Direction Mode"; break;
				case -5: strActiveMode = "Master Encoder Mode"; break;
				case -3: strActiveMode = "Current Mode"; break;
				case -2: strActiveMode = "Velocity Mode"; break;
				case -1: strActiveMode = "Position Mode"; break;
				case 1: strActiveMode = "Profile Position Mode"; break;
				case 3: strActiveMode = "Profile Velocity Mode"; break;
				case 6: strActiveMode = "Homing Mode"; break;
				case 7: strActiveMode = "Interpolated Position Mode"; break;
				default: strActiveMode = "Unknown Mode";
			}
		}
		else
		{
			StopTimer();
			ShowErrorInformation(m_dErrorCode);

			strActiveMode = "Unknown Mode";
		}
	}
	else
	{
		strActiveMode = "Unknown Mode";
	}

	if(oResult)
	{
		oResult = VCS_GetEnableState(m_KeyHandle,m_wNodeId,&oEnable,&m_dErrorCode);

		if(oResult)
		{
			SetCtrlAttribute(panel,PANEL_DeviceSettings,ATTR_DIMMED,oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonEnable,ATTR_DIMMED,oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonDisable,ATTR_DIMMED,!oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonMove,ATTR_DIMMED,!oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonHalt,ATTR_DIMMED,!oEnable);
		}
		else
		{
			StopTimer();
			ShowErrorInformation(m_dErrorCode);
			
			SetCtrlAttribute(panel,PANEL_DeviceSettings,ATTR_DIMMED,oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonEnable,ATTR_DIMMED,!oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonDisable,ATTR_DIMMED,oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonMove,ATTR_DIMMED,oEnable);
			SetCtrlAttribute(panel,PANEL_ButtonHalt,ATTR_DIMMED,oEnable);
		}
	}
	else
	{
		SetCtrlAttribute(panel,PANEL_DeviceSettings,ATTR_DIMMED,!oEnable);
		SetCtrlAttribute(panel,PANEL_ButtonEnable,ATTR_DIMMED,oEnable);
		SetCtrlAttribute(panel,PANEL_ButtonDisable,ATTR_DIMMED,oEnable);
		SetCtrlAttribute(panel,PANEL_ButtonMove,ATTR_DIMMED,oEnable);
		SetCtrlAttribute(panel,PANEL_ButtonHalt,ATTR_DIMMED,oEnable);
		SetCtrlAttribute(panel,PANEL_ButtonEnable,ATTR_DIMMED,oEnable);
	}

	if(oResult)
	{
		oResult = VCS_GetPositionIs(m_KeyHandle,m_wNodeId,&m_lPositionActual,&m_dErrorCode);
		if(!oResult)
		{
			StopTimer();
			ShowErrorInformation(m_dErrorCode);

			m_lPositionActual = 0;
			m_lPositionStart = 0;
		}
	}
	else
	{
		m_lPositionActual = 0;
		m_lPositionStart = 0;
	}

	SetCtrlVal(panel,PANEL_StateActiveMode,strActiveMode);
	SetCtrlVal(panel,PANEL_PositionActual,m_lPositionActual);
	SetCtrlVal(panel,PANEL_PositionStart,m_lPositionStart);

	SetCtrlVal(panel,PANEL_NodeId,m_wNodeId);

	return oResult;
}

/************************************************************************
If timer was started all 100ms this function will be called
************************************************************************/
int CVICALLBACK MonitorTimer (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_TIMER_TICK:

			if(!UpdateStatus(panel))
			{
				StopTimer();
				
			break;
	}
		}
	return 0;
}

/************************************************************************
Changes target position
************************************************************************/
int CVICALLBACK PositionTarget (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:		
			GetCtrlVal(panel,control,&m_lPositionTarget);
			break;
	}
	return 0;
}

/************************************************************************
Changes node id
************************************************************************/
int CVICALLBACK NodeId (int panel, int control, int event,
		void *callbackData, int eventData1, int eventData2)
{
	switch (event)
	{
		case EVENT_COMMIT:
			GetCtrlVal(panel,control,&m_wNodeId);
			break;
	}
	return 0;
}

File: EPOS/Microsoft Visual Basic/Example VB 2010 64-Bit/Demo_WinDLL.vb
Option Strict Off
Option Explicit On
Friend Class Demo_WinDLL
    Inherits System.Windows.Forms.Form
    Dim m_usNodeId As UShort
    Dim m_lErrorCode As UInteger
    Dim m_lImmediately As Boolean
    Dim m_lProfileAcceleration As UInteger
    Dim m_lProfileDeceleration As UInteger
    Dim m_lProfileVelocity As UInteger
    Dim m_KeyHandle As Integer
    Dim m_uMode As SByte
    Dim m_oInitialisation As Boolean
    Dim m_oUpdateActive As Boolean

    'Sets new communication settings via button 'Device Settings'
    Private Sub DeviceSettings_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles DeviceSettings.Click
        NodeID.Text = CStr(m_usNodeId)

        If (OpenDevice()) Then
            m_oInitialisation = True
            m_oUpdateActive = True
            Timer1.Enabled = True
        Else
            m_oInitialisation = StopTimer()
            m_oInitialisation = False
        End If

    End Sub

    'Sets device to disable state
    Private Sub Disable_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Disable.Click
        Dim oResult As Boolean

        NodeID.Text = CStr(m_usNodeId)

        oResult = VCS_SetDisableState(m_KeyHandle, m_usNodeId, m_lErrorCode)
        If Not oResult Then
            oResult = ShowErrorInformation(m_lErrorCode)
        End If

    End Sub

    'Sets device to enable state
    Private Sub Enable_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Enable.Click
        Dim oResult As Boolean

        NodeID.Text = CStr(m_usNodeId)

        oResult = VCS_SetEnableState(m_KeyHandle, m_usNodeId, m_lErrorCode)
        If Not oResult Then
            oResult = ShowErrorInformation(m_lErrorCode)
        End If

    End Sub

    Private Sub Form1_Load(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles MyBase.Load

        m_lImmediately = 0
        m_usNodeId = 1

        If (OpenDevice()) Then
            m_oInitialisation = True
            m_oUpdateActive = True
            Timer1.Enabled = True
        Else
            m_oInitialisation = False
            m_oUpdateActive = False
        End If

    End Sub

    Private Sub Form1_FormClosed(ByVal eventSender As System.Object, ByVal eventArgs As System.Windows.Forms.FormClosedEventArgs) Handles Me.FormClosed
        Dim oResult As Boolean

        oResult = RestoreEPOS()
    End Sub

    'Opens communication interface via the OpenDeviceDlg
    Private Function OpenDevice() As Boolean

        Dim oResult As Boolean
        Dim lValue As Integer

        m_KeyHandle = VCS_OpenDeviceDlg(m_lErrorCode)
        If m_KeyHandle Then
            'Deletes the error history
            oResult = VCS_ClearFault(m_KeyHandle, m_usNodeId, m_lErrorCode)
            If oResult Then
                'Reads Operational Mode
                oResult = VCS_GetOperationMode(m_KeyHandle, m_usNodeId, m_uMode, m_lErrorCode)
                If oResult Then
                    'Reads Position Profile Objects
                    oResult = VCS_GetPositionProfile(m_KeyHandle, m_usNodeId, m_lProfileVelocity, m_lProfileAcceleration, m_lProfileDeceleration, m_lErrorCode)
                    If oResult Then
                        'Write Opeational Mode (Profile Position Mode)
                        oResult = VCS_SetOperationMode(m_KeyHandle, m_usNodeId, 1, m_lErrorCode)
                        If oResult Then
                            'Write Position Profile Objects
                            oResult = VCS_SetPositionProfile(m_KeyHandle, m_usNodeId, 100, 1000, 1000, m_lErrorCode)
                            If oResult Then
                                'Reads the actual position
                                oResult = VCS_GetPositionIs(m_KeyHandle, m_usNodeId, lValue, m_lErrorCode)
                                If oResult Then
                                    PositionStart.Text = CStr(lValue)
                                End If
                            End If
                        End If
                    End If
                End If
            End If

            If Not oResult Then
                ShowErrorInformation(m_lErrorCode)
            End If

        End If

        OpenDevice = oResult
    End Function

    'Restores the old settings, stops the timer and delete the key handle
    Private Function RestoreEPOS() As Boolean
        Dim oResult As Boolean

        If m_oInitialisation Then
            oResult = VCS_SetDisableState(m_KeyHandle, m_usNodeId, m_lErrorCode)
            oResult = VCS_SetOperationMode(m_KeyHandle, m_usNodeId, m_uMode, m_lErrorCode)
            oResult = VCS_SetPositionProfile(m_KeyHandle, m_usNodeId, m_lProfileVelocity, m_lProfileAcceleration, m_lProfileDeceleration, m_lErrorCode)
        End If

        oResult = StopTimer()

        oResult = VCS_CloseAllDevices(m_lErrorCode)

        RestoreEPOS = oResult

    End Function

    'Updates the display
    Private Function UpdateStatus() As Boolean
        Dim oResult As Boolean
        Dim oTemp As Boolean
        Dim lEnable As Integer
        Dim lValue As Integer

        lEnable = 1
        oResult = m_oUpdateActive

        If (RadioRelative.Checked) Then
            MoveToPos.Text = "&Move Relative"
        Else
            MoveToPos.Text = "&Move Absolute"
        End If

        If oResult Then
            oResult = VCS_GetOperationMode(m_KeyHandle, m_usNodeId, m_uMode, m_lErrorCode)

            If (oResult) Then
                Select Case m_uMode
                    Case 1 : ActiveOperationMode.Text = "Profile Position Mode"
                    Case 3 : ActiveOperationMode.Text = "Profile Velocity Mode"
                    Case 6 : ActiveOperationMode.Text = "Homing Mode"
                    Case 7 : ActiveOperationMode.Text = "Interpolated Position Mode"
                    Case -6 : ActiveOperationMode.Text = "Step/Direction Mode"
                    Case -5 : ActiveOperationMode.Text = "Master Encoder Mode"
                    Case -3 : ActiveOperationMode.Text = "Current Mode"
                    Case -2 : ActiveOperationMode.Text = "Velocity Mode"
                    Case -1 : ActiveOperationMode.Text = "Position Mode"
                    Case Else : ActiveOperationMode.Text = "Unknown Mode"
                End Select
            Else
                oTemp = StopTimer()
                oTemp = ShowErrorInformation(m_lErrorCode)

                ActiveOperationMode.Text = "Unknown Mode"
            End If
        Else
            ActiveOperationMode.Text = "Unknown Mode"
        End If

        If (oResult) Then
            oResult = VCS_GetEnableState(m_KeyHandle, m_usNodeId, lEnable, m_lErrorCode)

            If oResult Then
                DeviceSettings.Enabled = Not lEnable
                Enable.Enabled = Not lEnable
                Disable.Enabled = lEnable
                MoveToPos.Enabled = lEnable
                Halt.Enabled = lEnable
            Else
                oTemp = StopTimer()
                oTemp = ShowErrorInformation(m_lErrorCode)

                DeviceSettings.Enabled = lEnable
                Enable.Enabled = lEnable
                Disable.Enabled = Not lEnable
                MoveToPos.Enabled = Not lEnable
                Halt.Enabled = Not lEnable
            End If
        Else
            DeviceSettings.Enabled = lEnable
            Enable.Enabled = Not lEnable
            Disable.Enabled = Not lEnable
            MoveToPos.Enabled = Not lEnable
            Halt.Enabled = Not lEnable
        End If

        If oResult Then
            oResult = VCS_GetPositionIs(m_KeyHandle, m_usNodeId, lValue, m_lErrorCode)
            If oResult Then
                PositionActual.Text = CStr(lValue)
            Else
                oTemp = StopTimer()
                oTemp = ShowErrorInformation(m_lErrorCode)

                PositionActual.Text = CStr(0)
            End If
        Else
            PositionActual.Text = CStr(0)
        End If

        UpdateStatus = oResult

    End Function

    'Stops the movement
    Private Sub Halt_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Halt.Click
        Dim oResult As Boolean

        NodeID.Text = CStr(m_usNodeId)

        oResult = VCS_HaltPositionMovement(m_KeyHandle, m_usNodeId, m_lErrorCode)
        If Not oResult Then
            oResult = ShowErrorInformation(m_lErrorCode)
        End If

    End Sub

    'Starts the movement
    Private Sub MoveToPos_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles MoveToPos.Click
        Dim oResult As Boolean
        Dim lValue As Integer

        NodeID.Text = CStr(m_usNodeId)

        oResult = VCS_GetPositionIs(m_KeyHandle, m_usNodeId, lValue, m_lErrorCode)
        If oResult Then
            PositionStart.Text = CStr(lValue)

            If RadioAbsolute.Checked Then
                oResult = VCS_MoveToPosition(m_KeyHandle, m_usNodeId, CInt(TargetPosition.Text), True, m_lImmediately, m_lErrorCode)
            Else
                oResult = VCS_MoveToPosition(m_KeyHandle, m_usNodeId, CInt(TargetPosition.Text), False, m_lImmediately, m_lErrorCode)
            End If

            If Not oResult Then
                oResult = ShowErrorInformation(m_lErrorCode)
            End If
        End If

    End Sub

    'Changes node id
    Private Sub NodeID_TextChanged(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles NodeID.TextChanged
        If (NodeID.Text = "") Then
        Else
            Timer1.Enabled = True
            m_usNodeId = CShort(NodeID.Text)
        End If

    End Sub

    Private Sub OK_Click(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles OK.Click
        Dim oResult As Boolean

        oResult = RestoreEPOS()
        End
    End Sub

    'closes the dialog
    Private Sub Timer1_Tick(ByVal eventSender As System.Object, ByVal eventArgs As System.EventArgs) Handles Timer1.Tick
        Dim oResult As Boolean

        oResult = UpdateStatus()
        If Not oResult Then
            oResult = StopTimer()
        End If

    End Sub

    'example to show how any object can be written or read
    Private Function SetObject() As Boolean
        Dim oResult As Boolean

        Dim index As UShort
        Dim subIndex As Byte
        Dim Data As Integer
        Dim NbOfBytesToWrite As Integer
        Dim NbOfBytesWritten As Integer

        'Min. Position Limit
        index = &H607DS
        subIndex = &H1S
        Data = 0
        NbOfBytesToWrite = 4
        'All information about index, subindex, data range and length are from document 'EPOS Firmware Specification'

        oResult = VCS_GetObject(m_KeyHandle, m_usNodeId, index, subIndex, Data, NbOfBytesToWrite, NbOfBytesWritten, m_lErrorCode)

        If (oResult) Then
            Data = -588
            ' Value from "EPOS Firmware Specification" in Value range
            oResult = VCS_SetObject(m_KeyHandle, m_usNodeId, index, subIndex, Data, NbOfBytesToWrite, NbOfBytesWritten, m_lErrorCode)
        End If

        If (oResult) Then
            oResult = VCS_Store(m_KeyHandle, m_usNodeId, m_lErrorCode)
        End If

        If Not oResult Then
            oResult = ShowErrorInformation(m_lErrorCode)
        End If

        SetObject = oResult

    End Function

    'Stops timer. Status will be displayed as disconnected
    Private Function StopTimer() As Boolean

        Timer1.Enabled = False

        m_oUpdateActive = False

        StopTimer = UpdateStatus()

    End Function

    'Shows a message box with error description of the input error code
    Private Function ShowErrorInformation(ByVal lErrorCode As Integer) As Boolean
        Const MAX_STR_SIZE As Short = 100

        Dim oResult As Boolean
        Dim pStrErrorInfo As New VB6.FixedLengthString(MAX_STR_SIZE)

        CodeString(lErrorCode)

        oResult = VCS_GetErrorInfo(lErrorCode, pStrErrorInfo.Value, MAX_STR_SIZE)

        If oResult Then
            MsgBox(pStrErrorInfo.Value, MsgBoxStyle.OkOnly, "Error")
        Else
            MsgBox("Error information can't be read!", MsgBoxStyle.OkOnly, "Error")
        End If

        ShowErrorInformation = oResult

    End Function

    'Example CodeString
    Private Function CodeString(ByVal lErrorCodeValue As Integer) As Boolean
        Const MAX_STR_SIZE As Short = 30

        Dim oResult As Boolean
        Dim ErrorInfo As New VB6.FixedLengthString(MAX_STR_SIZE)
        Dim RetErrorInfo As String

        'Execute Function
        oResult = VCS_GetErrorInfo(lErrorCodeValue, ErrorInfo.Value, MAX_STR_SIZE)

        'Convert null terminated String to VB String
        RetErrorInfo = Microsoft.VisualBasic.Left(ErrorInfo.Value, InStr(ErrorInfo.Value, Chr(0)) - 1)

        CodeString = oResult

    End Function

    'Example Code DataBuffer
    Private Function CodeDataBuffer() As Boolean
        Const MAX_STR_SIZE As Short = 19

        Dim oResult As Boolean

        Dim DataVectorBuffer(MAX_STR_SIZE) As Byte
        Dim ChannelNumber As Byte
        Dim VectorSize As Long

        'Execute Function
        ChannelNumber = 1
        VectorSize = MAX_STR_SIZE + 1
        oResult = VCS_ReadChannelDataVector(m_KeyHandle, m_usNodeId, ChannelNumber, DataVectorBuffer(0), VectorSize, m_lErrorCode)

        CodeDataBuffer = oResult

    End Function

End Class